From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Fri, 13 Mar 2026 07:45:42 +0100
Subject: ubuntu-insights: Add support for configuring whoopsie reports

Add a further switch to control whoopsie reports and adjust wording
---
 .../ubuntu-insights/gis-ubuntu-insights-page.blp   |  47 +++-
 .../ubuntu-insights/gis-ubuntu-insights-page.c     | 297 +++++++++++++++++++--
 2 files changed, 325 insertions(+), 19 deletions(-)

diff --git a/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.blp b/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.blp
index c50b770..58a98f2 100644
--- a/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.blp
+++ b/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.blp
@@ -5,10 +5,12 @@ template $GisUbuntuInsightsPage: $GisPage {
   Adw.PreferencesPage {
     Adw.PreferencesGroup {
       $GisPageHeader header {
-        margin-top: "24";
         icon-name: "ubuntu-insights";
         title: _("Help Improve Ubuntu");
+        subtitle: _("Help us improve Ubuntu by sharing your system and error reports with us.\nReports are sent anonymously.");
         show-icon: bind template.small-screen inverted;
+        spacing: 0;
+        use-accessible-title: true;
       }
     }
 
@@ -75,6 +77,49 @@ template $GisUbuntuInsightsPage: $GisPage {
       }
     }
 
+    Adw.PreferencesGroup whoopsie_group {
+      accessible-role: group;
+      visible: false;
+
+      accessibility {
+        labelled-by: [
+          header,
+        ];
+
+        described-by: [
+          header,
+        ];
+      }
+
+      Adw.SwitchRow whoopsie_switch {
+        title: _("Share _error reports with the Ubuntu team");
+        subtitle: _("This includes crash reports and related system logs");
+        use-underline: true;
+      }
+
+      CenterBox {
+        margin-top: 12;
+
+        [start]
+        Box {
+          halign: start;
+          valign: start;
+          spacing: 18;
+
+          LinkButton {
+            label: _("Learn more about error reporting");
+            uri: "https://documentation.ubuntu.com/project/contributors/debugging/apport/";
+            activate-link => $on_link_activated(template);
+
+            styles [
+              "caption",
+              "link-caption",
+            ]
+          }
+        }
+      }
+    }
+
     Adw.PreferencesGroup {
       $GisFocusableBin {
         vexpand: true;
diff --git a/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.c b/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.c
index 4bdab24..917847c 100644
--- a/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.c
+++ b/gnome-initial-setup/pages/ubuntu-insights/gis-ubuntu-insights-page.c
@@ -45,6 +45,8 @@ struct _GisUbuntuInsightsPage {
   GtkWidget *insights_switch;
   GtkWidget *header;
   GtkWidget *show_report_link;
+  GtkWidget *whoopsie_group;
+  GtkWidget *whoopsie_switch;
 
   AdwDialog *report_dialog;
   GtkWidget *report_stack;
@@ -52,6 +54,7 @@ struct _GisUbuntuInsightsPage {
   GtkWidget *report_scroll;
   GtkWidget *report_text_view;
 
+  GDBusProxy *whoopsie;
   GCancellable *cancellable;
   gchar *report;
 };
@@ -122,16 +125,203 @@ page_shown (GisPage *page)
   gtk_widget_grab_focus (self->header);
 }
 
+typedef enum {
+  ACCOUNT_TYPE_STANDARD      = 0,
+  ACCOUNT_TYPE_ADMINISTRATOR = 1,
+} AccountType;
+
+static char *
+get_user_object_path (GDBusConnection *connection,
+                      uid_t            uid,
+                      GCancellable    *cancellable,
+                      GError         **error)
+{
+  g_autofree char *path = NULL;
+  g_autoptr(GVariant) result = NULL;
+
+  result = g_dbus_connection_call_sync (connection,
+                                        "org.freedesktop.Accounts",
+                                        "/org/freedesktop/Accounts",
+                                        "org.freedesktop.Accounts",
+                                        "FindUserById",
+                                        g_variant_new ("(x)", (gint64) uid),
+                                        G_VARIANT_TYPE ("(o)"),
+                                        G_DBUS_CALL_FLAGS_NONE, -1,
+                                        cancellable, error);
+
+  if (!result)
+    return NULL;
+
+  g_variant_get (result, "(o)", &path);
+
+  return g_steal_pointer (&path);
+}
+
+gboolean
+user_is_admin (GCancellable  *cancellable,
+               GError       **error)
+{
+  g_autoptr(GDBusConnection) connection = NULL;
+  g_autoptr(GDBusProxy) proxy = NULL;
+  g_autoptr(GVariant) val = NULL;
+  g_autofree char *object_path = NULL;
+
+  connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, cancellable, error);
+  if (!connection)
+    return FALSE;
+
+  object_path = get_user_object_path (connection, getuid (), cancellable, error);
+  if (!object_path)
+    return FALSE;
+
+  proxy = g_dbus_proxy_new_sync (connection,
+                                 G_DBUS_PROXY_FLAGS_NONE, NULL,
+                                 "org.freedesktop.Accounts",
+                                 object_path,
+                                 "org.freedesktop.Accounts.User",
+                                 cancellable, error);
+
+  if (!proxy)
+    return FALSE;
+
+  val = g_dbus_proxy_get_cached_property (proxy, "AccountType");
+  if (!val)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to get AccountType property");
+      return FALSE;
+    }
+
+  return (AccountType) g_variant_get_int32 (val) == ACCOUNT_TYPE_ADMINISTRATOR;
+}
+
+static void
+whoopsie_properties_changed (GisUbuntuInsightsPage *page)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (G_DBUS_PROXY (page->whoopsie));
+  g_autoptr (GVariant) auto_report = NULL;
+
+  /* Bail out if dbus service has gone away */
+  if (G_UNLIKELY (!name_owner))
+    return;
+
+  auto_report = g_dbus_proxy_get_cached_property (G_DBUS_PROXY (page->whoopsie),
+                                                  "AutomaticallyReportCrashes");
+  gboolean enabled = g_variant_get_boolean (auto_report);
+  adw_switch_row_set_active (ADW_SWITCH_ROW (page->whoopsie_switch), enabled);
+}
+
+static void
+whoopsie_bus_owner_changed (GDBusProxy *proxy)
+{
+  g_autofree char *name_owner = g_dbus_proxy_get_name_owner (proxy);
+
+  if (!name_owner)
+    {
+      /* Trigger the service to start again, as it has just been stopped by a timeout
+       * don't bother on waiting for the call results.
+       */
+      g_dbus_proxy_call (proxy, "GetIdentifier", NULL, G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL, NULL);
+    }
+}
+
+static void
+on_whoopsie_preferences_proxy_ready (GObject      *source_object,
+                                     GAsyncResult *res,
+                                     gpointer      user_data)
+{
+  GisUbuntuInsightsPage *page = GIS_UBUNTU_INSIGHTS_PAGE (user_data);
+  g_autoptr(GError) error = NULL;
+
+  page->whoopsie = g_dbus_proxy_new_for_bus_finish (res, &error);
+  if (!page->whoopsie)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        return;
+
+      g_warning ("Failed to create WhoopsiePreferences proxy: %s",
+                 error->message);
+      return;
+    }
+
+  gtk_widget_set_visible (page->whoopsie_group, TRUE);
+  whoopsie_properties_changed (page);
+
+  g_signal_connect_object (page->whoopsie, "g-properties-changed",
+                           G_CALLBACK (whoopsie_properties_changed),
+                           page, G_CONNECT_SWAPPED);
+
+  g_signal_connect_object (page->whoopsie, "notify::g-name-owner",
+                           G_CALLBACK (whoopsie_bus_owner_changed),
+                           page, 0);
+}
+
+static void
+whoopsie_check_thread (GTask        *task,
+                       gpointer      source_object,
+                       gpointer      task_data,
+                       GCancellable *cancellable)
+{
+  g_autoptr(GError) error = NULL;
+  gboolean is_admin = user_is_admin (cancellable, &error);
+
+  if (error)
+    {
+      g_task_return_error (task, g_steal_pointer (&error));
+      return;
+    }
+
+  g_task_return_boolean (task, is_admin);
+}
+
+static void
+on_whoopsie_availability_check_done (GObject      *source_object,
+                                     GAsyncResult *res,
+                                     gpointer      task_data)
+{
+  GisUbuntuInsightsPage *page = GIS_UBUNTU_INSIGHTS_PAGE (source_object);
+  g_autoptr(GError) error = NULL;
+  gboolean is_admin;
+
+  gis_page_set_complete (GIS_PAGE (page), TRUE);
+
+  is_admin = g_task_propagate_boolean (G_TASK (res), &error);
+  if (error)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        g_warning ("Error checking admin status: %s", error->message);
+
+      return;
+    }
+
+  if (!is_admin)
+    {
+      g_message ("Not an administrator account, skipping report collection");
+      return;
+    }
+
+  g_dbus_proxy_new_for_bus (G_BUS_TYPE_SYSTEM, G_DBUS_PROXY_FLAGS_NONE, NULL,
+                            "com.ubuntu.WhoopsiePreferences",
+                            "/com/ubuntu/WhoopsiePreferences",
+                            "com.ubuntu.WhoopsiePreferences",
+                            page->cancellable,
+                            on_whoopsie_preferences_proxy_ready,
+                            page);
+}
+
 static void
 gis_ubuntu_insights_page_constructed (GObject *object)
 {
   GisUbuntuInsightsPage *page = GIS_UBUNTU_INSIGHTS_PAGE (object);
+  g_autoptr(GTask) task = NULL;
 
   G_OBJECT_CLASS (gis_ubuntu_insights_page_parent_class)->constructed (object);
 
   page->cancellable = g_cancellable_new ();
 
-  gis_page_set_complete (GIS_PAGE (page), TRUE);
+  task = g_task_new (page, page->cancellable, on_whoopsie_availability_check_done, page);
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_run_in_thread (task, whoopsie_check_thread);
 }
 
 static void
@@ -141,6 +331,7 @@ gis_ubuntu_insights_page_finalize (GObject *object)
 
   g_cancellable_cancel (page->cancellable);
   g_clear_object (&page->cancellable);
+  g_clear_object (&page->whoopsie);
 
   g_free (page->report);
 
@@ -310,11 +501,15 @@ typedef struct {
   gboolean consent;
   gboolean collect;
   gchar *report;
+
+  gboolean whoopsie_reporting;
+  GDBusProxy *whoopsie;
 } InsightsApplyThreadData;
 
 static void
 ubuntu_insights_apply_thread_data_free (InsightsApplyThreadData *data)
 {
+  g_clear_object (&data->whoopsie);
   g_free (data->report);
   g_free (data);
 }
@@ -328,6 +523,17 @@ on_ubuntu_insights_apply_thread (GTask        *task,
   InsightsApplyThreadData *data = task_data;
   g_autofree char *error_msg = NULL;
 
+  if (data->whoopsie)
+    {
+      g_dbus_proxy_call (data->whoopsie, "SetReportCrashes",
+                         g_variant_new ("(b)", data->whoopsie_reporting),
+                         G_DBUS_CALL_FLAGS_NONE, -1, cancellable, NULL, NULL);
+
+      g_dbus_proxy_call (data->whoopsie, "SetAutomaticallyReportCrashes",
+                         g_variant_new ("(b)", data->whoopsie_reporting),
+                         G_DBUS_CALL_FLAGS_NONE, -1, cancellable, NULL, NULL);
+    }
+
   for (size_t i = 0; i < G_N_ELEMENTS (INSIGHT_SOURCES); i++) {
     error_msg = insights_set_consent_state (NULL, INSIGHT_SOURCES[i].source,
                                             data->consent);
@@ -418,6 +624,9 @@ gis_ubuntu_insights_page_apply (GisPage *gis_page, GCancellable *cancellable)
   data->collect = gis_driver_get_mode (gis_page->driver) != GIS_DRIVER_MODE_UPGRADE;
   data->report = g_steal_pointer (&page->report);
 
+  data->whoopsie_reporting = adw_switch_row_get_active (ADW_SWITCH_ROW (page->whoopsie_switch));
+  g_set_object (&data->whoopsie, page->whoopsie);
+
   task = g_task_new (page, cancellable, on_ubuntu_insights_apply_callback, NULL);
   g_task_set_task_data (task, g_steal_pointer (&data),
                         (GDestroyNotify) ubuntu_insights_apply_thread_data_free);
@@ -431,21 +640,12 @@ static void
 gis_ubuntu_insights_page_locale_changed (GisPage *gis_page)
 {
   GisUbuntuInsightsPage *page = GIS_UBUNTU_INSIGHTS_PAGE (gis_page);
-  g_autofree gchar *description_text = NULL;
-
-  gis_page_set_title (GIS_PAGE (page), _("Help Improve Ubuntu"));
-
-  description_text = g_strdup_printf (_("Help us improve Ubuntu by sharing your system data with us. "
-                                        "This includes things like your machine model, "
-                                        "language, and timezone."));
 
   if (gis_driver_get_mode (gis_page->driver) != GIS_DRIVER_MODE_UPGRADE)
     {
       gtk_button_set_label (GTK_BUTTON (page->show_report_link),
                             _("Show First Report"));
     }
-
-  g_object_set (page->header, "subtitle", description_text, NULL);
 }
 
 static gboolean
@@ -467,6 +667,8 @@ gis_ubuntu_insights_page_class_init (GisUbuntuInsightsPageClass *klass)
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, header);
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, show_report_link);
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, insights_switch);
+  gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, whoopsie_group);
+  gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, whoopsie_switch);
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, report_dialog);
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, report_stack);
   gtk_widget_class_bind_template_child (GTK_WIDGET_CLASS (klass), GisUbuntuInsightsPage, report_loading_spinner);
@@ -485,18 +687,66 @@ gis_ubuntu_insights_page_class_init (GisUbuntuInsightsPageClass *klass)
   gis_add_style_from_resource ("/org/gnome/initial-setup/gis-ubuntu-insights-page.css");
 }
 
-gboolean
+insights_consent_state
 insight_needs_consent (void)
 {
+  gboolean all_granted = TRUE;
+
   for (size_t i = 0; i < G_N_ELEMENTS (INSIGHT_SOURCES); i++)
     {
       const char *source = INSIGHT_SOURCES[i].source;
+      insights_consent_state state = insights_get_consent_state (NULL, source);
+
+      if (state == INSIGHTS_CONSENT_UNKNOWN)
+        return INSIGHTS_CONSENT_UNKNOWN;
+
+      if (state != INSIGHTS_CONSENT_TRUE)
+        all_granted = FALSE;
+    }
+
+  return all_granted ? INSIGHTS_CONSENT_TRUE : INSIGHTS_CONSENT_FALSE;
+}
+
+static gboolean
+whoopsie_is_available (void)
+{
+  g_autoptr(GDBusConnection) connection = NULL;
+  g_autoptr(GVariant) result = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree const char **names = NULL;
 
-      if (insights_get_consent_state (NULL, source) == INSIGHTS_CONSENT_UNKNOWN)
-        return TRUE;
+  if (!user_is_admin (NULL, &error))
+    {
+      if (error)
+        g_warning ("Failed to check admin status: %s", error->message);
+      return FALSE;
     }
 
-  return FALSE;
+  connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error);
+  if (!connection)
+    {
+      g_warning ("Failed to connect to system bus: %s", error->message);
+      return FALSE;
+    }
+
+  result = g_dbus_connection_call_sync (connection,
+                                        "org.freedesktop.DBus",
+                                        "/org/freedesktop/DBus",
+                                        "org.freedesktop.DBus",
+                                        "ListActivatableNames",
+                                        NULL,
+                                        G_VARIANT_TYPE ("(as)"),
+                                        G_DBUS_CALL_FLAGS_NONE, -1,
+                                        NULL, &error);
+  if (!result)
+    {
+      g_warning ("Failed to list activatable D-Bus names: %s", error->message);
+      return FALSE;
+    }
+
+  g_variant_get (result, "(^a&s)", &names);
+
+  return g_strv_contains (names, "com.ubuntu.WhoopsiePreferences");
 }
 
 static void
@@ -511,15 +761,26 @@ gis_ubuntu_insights_page_init (GisUbuntuInsightsPage *page)
 GisPage *
 gis_prepare_ubuntu_insights_page (GisDriver *driver)
 {
+  GisUbuntuInsightsPage *page;
+  insights_consent_state consent;
+
+  consent = insight_needs_consent ();
+
   if (gis_driver_get_mode (driver) == GIS_DRIVER_MODE_UPGRADE &&
-      !insight_needs_consent () &&
-      !g_getenv ("GIS_SHOW_ALL_PAGES"))
+      consent != INSIGHTS_CONSENT_UNKNOWN &&
+      !g_getenv ("GIS_SHOW_ALL_PAGES") &&
+      !whoopsie_is_available ())
     {
-      g_message ("Ubuntu Insights consent has already been set. Skipping Ubuntu Insights page.");
+      g_message ("Ubuntu Insights consent has already been set and whoopsie is not available. "
+                 "Skipping Ubuntu Insights page.");
       return NULL;
     }
 
-  return g_object_new (GIS_TYPE_UBUNTU_INSIGHTS_PAGE,
+  page = g_object_new (GIS_TYPE_UBUNTU_INSIGHTS_PAGE,
                        "driver", driver,
                        NULL);
+  adw_switch_row_set_active (ADW_SWITCH_ROW (page->insights_switch),
+                             consent == INSIGHTS_CONSENT_TRUE);
+
+  return GIS_PAGE (page);
 }
